kqueen/engines/openstack_kubespray.py

```diff
@@ -561,38 +568,60 @@ def __init__(self, stack_name, *, os_kwargs, cluster, extra_ssh_key):
         self.os_kwargs = os_kwargs
+        self.meta = {}
+
+    def validate_cluster_metadata(self):
+        def validate_ip(address):
+            address = address.strip()
+            # Raises ValueError if IP address is not valid
+            ipaddress.ip_address(address)
+            return address
+
+        mc = self.cluster.metadata["master_count"]
+        if mc % 2 == 0 or mc == 1:
+            raise ValueError("Master node count must be an odd number at least 3 or greater")
+        self.meta["master_count"] = mc
+        self.meta["slave_count"] = self.cluster.metadata["slave_count"]
+
+        self.meta["dns"] = [validate_ip(ip) for ip in
+                            self.cluster.metadata.get("dns_nameservers", []).split(",")]
+
+        self.meta["ext_net"] = self.c.get_network(self.cluster.metadata["floating_network"])
+        if self.meta["ext_net"] is None:
+            raise ValueError("External network '%s' is not found" % self.cluster.metadata["floating_network"])
+        self.meta["image"] = self.c.get_image(self.cluster.metadata["image_name"])
+        if self.meta["image"] is None:
+            raise ValueError("Image '%s' is not found" % self.cluster.metadata["image_name"])
+        self.meta["flavor"] = self.c.get_flavor(self.cluster.metadata["flavor"])
+        if self.meta["flavor"] is None:
+            raise ValueError("Flavor '%s' is not found" % self.cluster.metadata["flavor"])
+
+        azone = self.cluster.metadata.get("availability_zone")
+        if azone and azone not in self.c.list_availability_zone_names():
+            raise ValueError("Availability zone '{}' is not found".format(self.cluster.metadata["availability_zone"]))
+        else:
+            self.meta["azone"] = azone
 
     def provision(self):
-        master_count = self.cluster.metadata["master_count"]
-        slave_count = self.cluster.metadata["slave_count"]
-        dns = self.cluster.metadata["dns_nameservers"].split(",")
-        ext_net = self.c.get_network(self.cluster.metadata["floating_network"])
-        if ext_net is None:
-            raise Exception("External network %s not found" % self.cluster.metadata["floating_network"])
-        image = self.c.get_image(self.cluster.metadata["image_name"])
-        if image is None:
-            raise Exception("Image %s not found" % self.cluster.metadata["image_name"])
-        flavor = self.c.get_flavor(self.cluster.metadata["flavor"])
-        if flavor is None:
-            raise Exception("Flavor %s not found" % self.cluster.metadata["flavor"])
+
         resources = {
             "masters": [],
             "slaves": [],
         }
         network = self.c.create_network(self.stack_name)
         subnet = self.c.create_subnet(network, cidr="10.1.0.0/16",
                                       subnet_name=self.stack_name,
-                                      dns_nameservers=dns)
+                                      dns_nameservers=self.meta['dns'])
```

Comment:
maybe get() ?

Reply:
Some fields are required and some are not, let me also send 400 if required field is not provided

Reply:
okok